core/drivers/atmel_trng.c

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2021 Microchip
+ */
+
+#include <assert.h>
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
+#include <io.h>
+#include <kernel/boot.h>
+#include <libfdt.h>
+#include <matrix.h>
+#include <platform_config.h>
+#include <rng_support.h>
+#include <string.h>
+#include <tee/tee_cryp_utl.h>
+
+/* Registers */
+#define TRNG_CTRL		0x0
+#define TRNG_CTRL_WAKEY_OFFSET	8
+#define TRNG_CTRL_WAKEY_VALUE	0x524E47
+
+#define TRNG_IER		0x10
+#define TRNG_ISR		0x1C
+#define TRNG_ODATA		0x50
+
+static vaddr_t trng_base;
+static uint32_t trng_last_val;
+static uint8_t trng_valid_bytes;
+
+static uint32_t atmel_trng_read32(void)
+{
+	while (!io_read32(trng_base + TRNG_ISR))
+		;
+
+	return io_read32(trng_base + TRNG_ODATA);
+}
+
+TEE_Result crypto_rng_read(void *buf, size_t len)
+{
+	uint8_t *rngbuf = buf;
+	uint32_t val = 0;
+	uint8_t len_to_copy = 0;
+
+	assert(buf);
+	assert(trng_base);
+
+	while (len) {
+		val = atmel_trng_read32();
+		len_to_copy = MIN(len, sizeof(uint32_t));
+		memcpy(rngbuf, &val, len_to_copy);
+		rngbuf += len_to_copy;
+		len -= len_to_copy;
+	}
+
+	return TEE_SUCCESS;
+}
+
+uint8_t hw_get_random_byte(void)
+{
+	uint8_t data = 0;
+
+	assert(trng_base);
+
+	/*
+	 * The TRNG generates a whole 32 bits word every 84 cycles. To avoid
+	 * discarding 3 bytes at each request, request 4 bytes of random data
+	 * and return only 1 at each request until there is no more bytes in the
+	 * trng_valid_bytes "cache".
+	 */
+	if (trng_valid_bytes == 0) {
+		trng_last_val = atmel_trng_read32();
+		trng_valid_bytes = sizeof(uint32_t);
+	}
+
+	data = trng_last_val & 0xFF;
+	trng_last_val >>= 8;
```

Comment:
It looks like this global variable needs to be protected from concurrent access.

Reply:
Most probably yes, `trng_last_val` & `trng_valid_bytes` needs to be protected.